### PR TITLE
Replace bundle string handling with parse + traverse

### PR DESCRIPTION
**Pull request by tptee**
I ran into this when integrating with `webpack-dashboard`:
https://github.com/webpack/webpack/blob/93ac8e9c3699bf704068eaccaceec57b3dd45a14/lib/Template.js#L118

Turns out that Webpack bundles can contain either an array of modules or an object of id keys/module values. The object version seems rare, but I caught one in a simple lodash + entrypoint bundle, and inspectpack can't currently understand it.

This PR parses the bundles with `babylon`, traverses the AST, and creates `Code` objects from AST data. It parses both the array and object forms of bundles. Most string manipulation is gone, save for a bit of trickery in extracting data from module header comments. I figure this approach will make inspectpack more resistant to changes in the Webpack bundle format in the future.

I also modified the `Code` API a little, since we know more about the type of `Code` objects a priori during traversal. I've updated its usage in `models/bundle.js` and `actions/sizes.js`, but need to update the remainder of the action files.

@ryan-roemer Do you have example bundles that trigger non-empty cases for duplicates/files/parse/pattern/versions? I added some super-basic smoke tests for each action, but they won't be meaningful unless they're run against a bundle with positive results.

### CODE REVIEW

File: lib/models/bundle.js [hunk 1]
```diff
@@ -234,9 +44,9 @@ Bundle.prototype.validate = function () {
 
   codes.forEach(function (code) {
     // Single number refs -> real code.
-    if (code.ref !== null) {
-      var singleRef = codes[code.ref];
-      if (!singleRef.isCode) {
+    if (code.isSingleRef()) {
+      var singleRef = _.find({ id: code.singleRef })(codes);
```
Comment (tptee) [inline lib/models/bundle.js]: With object expressions, it's no longer safe to rely on the array indices of `codes`. We could consider converting the `codes` array into a hash of id keys/code pairs if we want O(1) lookups again.

**Comment by tptee** [comment]
Some rough edges left, but probably worth an initial pass!

File: lib/actions/duplicates.js [hunk 2]
```diff
@@ -38,7 +38,8 @@ var addSizes = function (codes, opts) {
     // Update summary items.
     _.each.convert({ cap: false })(function (obj, idx) {
       // Mutate sizes.
-      var codeSrc = toParseable(codes[idx].code);
+      // TODO: fix this lookup
```
Comment (divmain) [inline lib/actions/duplicates.js]: Is this a pre-merge placeholder, or something that needs to be ticketed out?

File: lib/utils/parser.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+var getModuleType = function (node) {
+  // A straight code reference.
+  //
+  // ```js
+  //
+  // /*!**************************!*\
+  //   !*** ../foo/awesomez.js ***!
+  //   \**************************/
+  // /***/ function(module, exports, __webpack_require__) {   <-- Code (null) L4
+  // ```
```
Comment (divmain) [inline lib/utils/parser.js]: These comments are super helpful!

File: lib/models/bundle.js (hunk 1 above)
Comment (ryan-roemer) [inline lib/models/bundle.js]: Do we have an example of a bundle in the object format? (I haven't finished skimming this PR yet).

File: package.json [hunk 4]
```diff
@@ -17,9 +17,10 @@
     "inspectpack": "bin/inspectpack.js"
   },
   "scripts": {
-    "lint-server": "eslint --color -c .eslintrc-server lib bin",
+    "lint-server": "eslint --color -c .eslintrc lib bin",
```
Comment (ryan-roemer) [inline package.json]: 1. `-c .eslintrc` can be removed as it's default
2. Are we missing `lint-test` task and aggregation in `lint`?

**Comment by ryan-roemer** [comment]
Looking great so far! Let me know when this hits complete and what specific types of examples you need for the smoke tests.

File: lib/models/bundle.js (hunk 1 above)
Comment (tptee) [inline lib/models/bundle.js]: Sure! Check out https://github.com/FormidableLabs/inspectpack/blob/4909a7644cf29a8f0c3df8498d10b4125eba2467/test/fixtures/basic-lodash-object-expression.js

Definitely the first time I've run into this...

File: lib/actions/base.js [hunk 5]
```diff
@@ -39,9 +39,11 @@ Base.prototype.display = function (data) {
   var opts = this.opts;
   var format = opts.format;
 
-  if (format === "json") {
+  if (format === "object") {
+    return data;
+  } else if (format === "json") {
     return JSON.stringify(data, null, JSON_SPACES);
-  } else if (format === "text") {
+  } if (format === "text") {
```
Comment (ryan-roemer) [inline lib/actions/base.js]: NIT: `} if` needs a newline to split things up. Should lint catch this for us?

**Comment by tptee** [comment]
I've neutered the `versions` smoke test for now, since doing it right requires with some nasty `postinstall`ing or heavy mocking 😬

File: lib/actions/base.js (hunk 5 above)
Comment (tptee) [inline lib/actions/base.js]: Replaced with a switch statement 😄 Surprised that I couldn't find a lint rule that covers this, though. It's pretty devious!

**Comment by tptee** [comment]
Should be ready for a final peek!

**Comment by ryan-roemer** [comment]
@tptee -- The fixtures are huge and I'd rather they not be part of this git repo. Can we move them to a separate published packaged? Like `inspectpack-test-fixtures` or something?

**Comment by tptee** [comment]
Done!

**Pull request merged**